### PR TITLE
Moved IEnhancedRandom Methods to Extensions Methods, Misc Reorganization

### PR DESCRIPTION
**Pull request by Chris3606**
# Changes
- Split the elements of `IEnhancedRandom.cs` out into their own files
    - `BitExtensions` -> `BitExtensions.cs`
    - `IEnhancedRandom` -> `Generators/IEnhancedRandom.cs`
    - `AbstractRandom` -> `Generators/AbstractRandom.cs`
- Moved `IEnhancedRandom` and `AbstractRandom` to the `ShaiRandom.Generators` namespace
- Removed some methods from `IEnhancedRandom` and `AbstractRandom`, and instead implemented them as `IEnhancedRandom` extension methods
    - Located in new file: `Generators/IEnhancedRandomExtensionMethods.cs`
    
# End State/Goal
This PR focuses on two areas:
1. Creating a more consistent and manageable organizational structure
2. Attempting to limit the amount of repetitive work someone has to do to implement `IEnhancedRandom` without using `AbstractRandom`

Since `EnhancedRandom.cs` was getting to be a huge file, I figured it made sense to pull it out into components, especially as I would have ended up adding another class to it.  I was unsure on what namespace `IEnhancedRandom` should actually belong in, however since `IEnhancedDistribution` was in the `Distributions` namespace, it seemed that `IEnhancedRandom` in the `Generators` namespace made the most sense to me.

As for moving things to extension methods, I tried to adhere to the following guidelines:
1. If a method had to do with the actual method of generating a primitive (int, byte, bool, or otherwise), serialization, or state setting/initialization, I left in in `IEnhancedRandom`.
2. If a method instead took a primitive that was generated and did some useful transformation or check on it, I moved it to an extension method.  A non-inclusive list of things moved and why:
    - `NextBool(float)`: Doesn't have to do with how to generate a random number; this method is simply an if-check around the result of a `NextFloat` call
    - `NextSign()`: Doesn't directly have to do with how to generate a number; rather takes a number generated adheres to the `IEnhancedRandom` contract, and generates a 1 or -1.
    - `MinIntOf`, `MaxIntOf`, and similar: These don't have to do with how to generate numbers; simply reporting results
    - `NextNormal`, `NextTriangular`: This is a method of taking numbers generated via the `IEnhancedRandom` contract for generating primitive numbers, and applying a particular distribution.

In short: methods that have to do with, or in some implementation could reasonably have to do with, _how_ to generate a random number from a state stayed on the interface.  Wrappers around such methods moved to extensions.

The idea is that this will prevent someone from having to re-implement these methods even if they need to implement `IEnhancedRandom` from scratch without the help of `AbstractRandom`.  A good use case for this is `KnownSeriesRandom`, which is bugged due to its usage of `new` right now; and I would advocate that the best fix for that IMO would be to have `KnownSeriesRandom` implement `IEnhancedRandom` instead of inheriting from `AbstractRandom`.  In the previous implementation, this would require `KnownSeriesRandom` to implement  22 methods in addition to what it does right now; most of which would be completely non-sensical for `KnownSeriesRandom` to be implementing.  After this PR, it would only require manually implementing `SetWith()`.  In fact, `SetWith` could technically be implemented as an extension as well; but I didn't do so in an effort to adhere strictly to my above criteria for an initial pass.

Please feel free to let me know what you think; if you don't like the extension methods or think certain things should or shouldn't be moved to extension methods, feel free to discuss and I can adapt the PR.  This is meant largely as a demonstration of what an extension method implementation could look like, as well as a starting point for a discussion on what to move.

### CODE REVIEW

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        /// <br/>
+        /// This can be used both as an optimization for generating Gaussian random values, and as a way of generating
+        /// Gaussian values that match a pattern present in the inputs (which you could have by using a sub-random sequence
+        /// as the input, such as those produced by a van der Corput, Halton, Sobol or R2 sequence). Most methods of generating
+        /// Gaussian values (e.g. Box-Muller and Marsaglia polar) do not have any way to preserve a particular pattern.
+        /// </remarks>
+        /// <param name="rng" />
+        /// <param name="d">should be between 0 and 1, exclusive, but other values are tolerated</param>
+        /// <returns>a normal-distributed double centered on 0.0; all results will be between -38.5 and 38.5, both inclusive</returns>
+        public static double Probit(this IEnhancedRandom rng, double d)
```
Comment (tommyettinger) [inline ShaiRandom/Generators/IEnhancedRandomExtensions.cs]: This is sort of an outlier; Probit(double) was static before (it was used in NextGaussian()), and doesn't need to take an IEnhancedRandom, just a double. I'm not sure where it should go.

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+        public static void Shuffle<T>(this IEnhancedRandom rng, T[] items)
+            => rng.Shuffle(items, 0, items.Length);
+
+        /// <summary>
+        /// Shuffles the given IList in-place pseudo-randomly, using the Fisher-Yates (also called Knuth) shuffle algorithm.
+        /// </summary>
+        /// <param name="rng" />
+        /// <param name="items">an IList; must be non-null but may contain null items</param>
+        public static void Shuffle<T>(this IEnhancedRandom rng, IList<T> items)
+            => rng.Shuffle(items, 0, items.Count);
```
Comment (tommyettinger) [inline ShaiRandom/Generators/IEnhancedRandomExtensions.cs:286]: These look good; is the plan to leave Span shuffling to a later time?

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+        /// Gets a normally-distributed (Gaussian) double, with a the specified mean (default 0.0) and standard deviation (default 1.0).
+        /// If the standard deviation is 1.0 and the mean is 0.0, then this can produce results between -8.209536145151493 and 8.209536145151493 (both extremely rarely).
+        /// </summary>
+        /// <param name="rng" />
+        /// <param name="mean">Mean for normal distribution.</param>
+        /// <param name="stdDev">Standard deviation for normal distribution.</param>
+        /// <returns>A double from the normal distribution with the specified mean (default 0.0) and standard deviation (default 1.0).</returns>
+        public static double NextNormal(this IEnhancedRandom rng, double mean = 0.0, double stdDev = 1.0)
+        {
+            return rng.Probit(rng.NextExclusiveDouble()) * stdDev + mean;
```
Comment (tommyettinger) [inline ShaiRandom/Generators/IEnhancedRandomExtensions.cs]: This probably just needs to be a call to a static Probit(double) method.

File: ShaiRandom/Generators/AbstractRandom.cs [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+            int i = 0;
+            for (; i < myCount && i < otherCount; i++)
+            {
+                SetSelectedState(i, other.SelectState(i));
+            }
+            for (; i < myCount; i++)
+            {
+                SetSelectedState(i, 0xFFFFFFFFFFFFFFFFUL);
+            }
+        }
```
Comment (tommyettinger) [inline ShaiRandom/Generators/AbstractRandom.cs]: SetWith() should probably be moved to an extension method to avoid some extra effort while implementing.

**Review by tommyettinger — CHANGES_REQUESTED** [review]
Only some minor stuff here; I agree that SetWith should be an extension method, and the other extension methods make sense where they are (except Probit(), which may belong in AbstractRandom next to the static AreEqual() method, or both may belong in some utility class). It definitely will help to have the interface easier to work with, since work in the future will probably go towards special-purpose generators that could certainly implement IEnhancedRandom directly.

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs (hunk 2 above)
Comment (Chris3606) [reply ShaiRandom/Generators/IEnhancedRandomExtensions.cs:286]: Yeah, separate PR; there's an argument for adding a variant of `NextBytes` that deals with Span too so I wanted to leave Span to a separate time.

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs (hunk 1 above)
Comment (Chris3606) [reply ShaiRandom/Generators/IEnhancedRandomExtensions.cs]: Moved to a static method in `MathUtil`.

File: ShaiRandom/Generators/IEnhancedRandomExtensions.cs (hunk 3 above)
Comment (Chris3606) [reply ShaiRandom/Generators/IEnhancedRandomExtensions.cs]: Done.

File: ShaiRandom/Generators/AbstractRandom.cs (hunk 4 above)
Comment (Chris3606) [reply ShaiRandom/Generators/AbstractRandom.cs]: Moved to extension method.

**Comment by Chris3606** [comment]
I believe I've covered everything mentioned; let me know if I missed anything!

**Review by tommyettinger — APPROVED** [review]
This looks great. I'm a little confused by the warnings on KnownSeriesRandom, but I'll merge and try to fix it myself. Happy New Year!

**Pull request merged**